tblite: init at 0.3.0

sheepforce opened the pull request:
###### Description of changes

Part of the continous effort to upstream packages from [NixOS-QChem](https://github.com/Nix-QChem/NixOS-QChem).
This introduces the tblite library and its ecosystem. Tblite is the foundation of some tight binding quantum chemistry packages, a prerequisite for the update of NWchem and the addition of xtb and dftb+

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sheepforce committed:
json-fortran: init at 8.3.0

sheepforce committed:
mctc-lib: init at 0.3.1


mctc-lib: add default.nix

sheepforce committed:
mstore: init at 0.2.0

sheepforce committed:
multicharge: init at 0.2.0

sheepforce committed:
test-drive: init at 0.4.0

sheepforce committed:
dftd4: init at 3.5.0

markuskowa commented on pkgs/development/libraries/science/chemistry/tblite/python.nix:
```diff
@@ -0,0 +1,33 @@
+}:
+
+buildPythonPackage rec {
+  pname = tblite.pname;
+  version = tblite.version;
+
+  src = tblite.src;
```
```suggestion
  inherit (tblite) pname version src meta;
```

markuskowa commented on pkgs/development/libraries/science/chemistry/tblite/python.nix:
```diff
@@ -0,0 +1,33 @@
+
+  nativeBuildInputs = [ meson ninja pkg-config ];
+
+  buildInputs = [ tblite ];
```
You use the sources from `tblite` and the package as build inputs at the same time? Does that cause ofborg to end up with an infinite recursion error?

sheepforce replied:
I'm afraid I do not understand completely what you mean by this :grimacing:

markuskowa replied:
Do we need `buildInputs = [ tblite ];` here? We already use it as a source (` src=tblite.src`).

sheepforce replied:
Ah yes, we do. Building the python interface is a three step procedure:

  1. Build the Fortran lib (the original tblite derivation)
  2. Build C-Interface (`configurePhase` in `python.nix`)
  3. Build the Python package around the C-Interface

It explicitly requires to have completed step 1 first and install tblite with tblite.pc files. Thus, yes we need it.

sheepforce replied:
Now I see the recursion error of OfBorg. Didn't noticed at all, first. The issue is much simpler than this. In `python-packages.nix` i have to pass `tblite = pkgs.tblite` and `meson = pkgs.meson` instead of `super.meson`  and `super.tblite`. `super` is the python package set. Is fixed now

sheepforce committed:
tblite: init at 0.3.0


tblite: inherit attributes of python derivation from parent


tblite: pass meson and tblite explicitly from pkgs to callPackage

markuskowa commented:
`simple-dftd3` and `toml-f` fail to build on aarch64.

sheepforce committed:
simple-dftd3: init at 0.7.0


simple-dftd3: restrict evaluation to x86_64-linux

sheepforce committed:
toml-f: init at 0.3.1


toml-f: restrict evaluation to x86_64-linux
